### DISCUSSION TITLE
如何修复npm audit的漏洞

### DISCUSSION THREAD

**Discussion by haonanu** [umijs/umi#11000] [Q&A]
UMI3 执行npm audit后 found 10 vulnerabilities (2 low, 4 moderate, 3 high, 1 critical) in 2617 scanned packages.
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ node-fetch is vulnerable to Exposure of Sensitive            │
│               │ Information to an Unauthorized Actor                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ node-fetch                                                   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=2.6.7                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ umi-request                                                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ umi-request > isomorphic-fetch > node-fetch                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1091239                     │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ node-fetch is vulnerable to Exposure of Sensitive            │
│               │ Information to an Unauthorized Actor                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ node-fetch                                                   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=2.6.7                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @umijs/plugin-dva                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @umijs/plugin-dva > dva > isomorphic-fetch > node-fetch      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1091239                     │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ Prototype Pollution in immer                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ immer                                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=9.0.6                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @umijs/plugin-dva                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @umijs/plugin-dva > dva-immer > immer                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1089281                     │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ critical      │ Prototype Pollution in immer                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ immer                                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=9.0.6                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @umijs/plugin-dva                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @umijs/plugin-dva > dva-immer > immer                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1089656                     │
└───────────────┴──────────────────────────────────────────────────────────────┘

**Comment by fz6m** [comment]
umi 3 不维护了，建议升级到 umi 3

**Reply by haonanu** [reply]
我尝试升级到Umi4，但是只是修复了4 moderate，还是存在3 high vulnerabilities，我还需要怎么操作呢？

**Reply by haonanu** [reply]
我升级了node并且新建了一个umi4 simple项目，found 0 vulnerabilities.
但是当我创建一个Umi4 Antd pro 项目时，7 vulnerabilities (6 high, 1 critical),这些漏洞怎么修复呢？
# npm audit report

immer  <=9.0.5
Severity: critical
Prototype Pollution in immer - https://github.com/advisories/GHSA-c36v-fmgq-m8hx
Prototype Pollution in immer - https://github.com/advisories/GHSA-33f9-j839-rf8h
No fix available
node_modules/immer
  dva-immer  *
  Depends on vulnerable versions of dva
  Depends on vulnerable versions of immer
  node_modules/@umijs/plugins/node_modules/dva-immer
    @umijs/plugins  *
    Depends on vulnerable versions of dva-immer
    node_modules/@umijs/plugins
      @umijs/max  *
      Depends on vulnerable versions of @umijs/plugins
      node_modules/@umijs/max

node-fetch  <=2.6.6
Severity: high
The `size` option isn't honored after following a redirect in node-fetch - https://github.com/advisories/GHSA-w7rc-rwvf-8q5r
node-fetch is vulnerable to Exposure of Sensitive Information to an Unauthorized Actor - https://github.com/advisories/GHSA-r683-j2x4-v87g
fix available via `npm audit fix`
node_modules/isomorphic-fetch/node_modules/node-fetch
  isomorphic-fetch  2.0.0 - 2.2.1
  Depends on vulnerable versions of node-fetch
  node_modules/isomorphic-fetch
    dva  0.0.2 - 0.0.9 || >=1.2.0-beta1
    Depends on vulnerable versions of isomorphic-fetch
    node_modules/@umijs/plugins/node_modules/dva

7 vulnerabilities (6 high, 1 critical)

**Reply by fz6m** [reply]
如果需要 `0 vulnerabilities` ，就不要用 max 项目，目前前端项目依赖很复杂，有很多老旧的依赖不再更新或者很少更新了，所以大型项目都会有非常多的废弃、漏洞报告，如果有强诉求，就全部依赖都自己管理，不要用第三方库。

另外针对个别的依赖，可以考虑 yarn resolutions 或者 pnpm overrides 覆盖版本，patch commit 修复依赖等手法。

**Reply by haonanu** [reply]
好的，感谢回复  @fz6m ，我再试一试
